core/src/main/scala/com/microsoft/azure/synapse/ml/explainers/ICEExplainer.scala

```diff
@@ -0,0 +1,252 @@
+    val dependenceCol = DatasetExtensions.findUnusedColumnName("feature__dependence", predicted)
+
+    val explainTarget = extractTarget(predicted.schema, targetClassesColumn)
+    val result = predicted.withColumn(targetCol, explainTarget)
+
+    getKind.toLowerCase match {
+      case this.averageKind =>
+        // PDP output schema: 1 row * 1 col (pdp for the given feature: feature_value -> explanations)
+        result
+          .groupBy(feature)
+          .agg(Summarizer.mean(col(targetCol)).alias(dependenceCol))
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(dependenceCol)
+            ).alias(outputColName)
+          )
+
+      case this.individualKind =>
+        // ICE output schema: n rows * 2 cols (idCol + ice for the given feature: map(feature_value -> explanations))
+        result
+          .groupBy(idCol)
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(targetCol)
+            ).alias(outputColName)
+          )
+    }
+  }
+
+  def transform(ds: Dataset[_]): DataFrame = {
+    transformSchema(ds.schema)
+    val df = ds.toDF
+    val idCol = DatasetExtensions.findUnusedColumnName("idCol", df)
+    val targetClasses = DatasetExtensions.findUnusedColumnName("targetClasses", df)
+    val dfWithId = df
+      .withColumn(idCol, monotonically_increasing_id())
+      .withColumn(targetClasses, this.get(targetClassesCol).map(col).getOrElse(lit(getTargetClasses)))
+
+    // collect feature values for all features from original dataset - dfWithId
+    val (categoricalFeatures, numericFeatures) = (this.getCategoricalFeatures, this.getNumericFeatures)
+
+    val sampled: Dataset[Row] = this.get(numSamples).map(dfWithId.orderBy(rand()).limit).getOrElse(dfWithId).cache
+
+    val calcCategoricalFunc: ICECategoricalFeature => DataFrame = {
+      f: ICECategoricalFeature =>
+        val values = collectCategoricalValues(dfWithId, f)
+        calcDependence(sampled, idCol, targetClasses, f.name, values, f.getOutputColName)
+    }
+    val calcNumericFunc: ICENumericFeature => DataFrame = {
+      f: ICENumericFeature =>
+        val values = collectSplits(dfWithId, f)
+        calcDependence(sampled, idCol, targetClasses, f.name, values, f.getOutputColName)
+    }
+
+    val dependenceDfs = (categoricalFeatures map calcCategoricalFunc) ++ (numericFeatures map calcNumericFunc)
+
+    getKind.toLowerCase match {
+      case this.individualKind =>
+        dependenceDfs.reduceOption(_.join(_, Seq(idCol), "inner"))
+          .map(sampled.join(_, Seq(idCol), "inner").drop(idCol)).get
+      case this.averageKind =>
+        dependenceDfs.reduce(_ crossJoin _)
+    }
+  }
+
+  private def collectCategoricalValues[_](df: DataFrame, feature: ICECategoricalFeature): Array[_] = {
+    val featureCountCol = DatasetExtensions.findUnusedColumnName("__feature__count__", df)
+    val values = df
```

Comment:
You can remove `val values = ` and just make this `df.xyz()` statement the last statement in the function